src/fp32.rs

```diff
@@ -0,0 +1,296 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! Finite field arithmetic for any field GF(p) for which p < 2^32.
+
+use crate::fp::{hi32, lo32, MAX_ROOTS};
+
+/// This structure represents the parameters of a finite field GF(p) for which p < 2^32.
+///
+/// See also [`FieldParameters`](crate::fp::FieldParameters).
+#[derive(Debug, PartialEq, Eq)]
+pub(crate) struct FieldParameters32 {
+    /// The prime modulus `p`.
+    pub p: u32,
+    /// `base = 2^32`.
+    pub base: u128,
+    /// `mu = -p^(-1) mod base`.
+    pub mu: u32,
+    /// `r2 = (2^32)^2 mod p`.
+    pub r2: u32,
+    /// The `2^num_roots`-th -principal root of unity. This element is used to generate the
+    /// elements of `roots`.
+    pub g: u32,
+    /// The number of principal roots of unity in `roots`.
+    pub num_roots: usize,
+    /// Equal to `2^b - 1`, where `b` is the length of `p` in bits.
+    pub bit_mask: u32,
+    /// `roots[l]` is the `2^l`-th principal root of unity, i.e., `roots[l]` has order `2^l` in the
+    /// multiplicative group. `roots[0]` is equal to one by definition.
+    pub roots: [u32; MAX_ROOTS + 1],
+}
+
+impl FieldParameters32 {
+    /// Addition. The result will be in [0, p), so long as both x and y are as well.
+    #[inline(always)]
+    pub fn add(&self, x: u32, y: u32) -> u32 {
+        //   0,x
+        // + 0,y
+        // =====
+        //   c,z
+        let (z, carry) = x.overflowing_add(y);
+        //     c, z
+        // -   0, p
+        // ========
+        // b1,s1,s0
+        let (s0, b0) = z.overflowing_sub(self.p);
+        let (_s1, b1) = (carry as u32).overflowing_sub(b0 as u32);
+        // if b1 == 1: return z
+        // else:       return s0
+        let m = 0u32.wrapping_sub(b1 as u32);
+        (z & m) | (s0 & !m)
+    }
```

Comment:
This function appears to be identical to the method with the same name on `FieldParameters`. If we make `FieldParameters` generic, as recommended above, then we can deduplicate this. Perhaps there are other methods that can be de-duplicated in the same way? Definitely not `mul()`.

Reply:
To specialize the implementation of mul() you would do something like
```rust
impl FieldParameters<u32> {
   pub fn mul(blah) -> blah {
      blah
   }
}
```